core/src/edu/asu/c3simulator/simulation/business/Company.java

```diff
@@ -0,0 +1,374 @@
+	 * <p>
+	 * This method is intended to be called by Simulation and its agents, as part of a
+	 * market/economy simulation.
+	 * 
+	 * @param product
+	 *            The product to be bought/sold
+	 * @param quantity
+	 *            The quantity to buy/sell
+	 * @return True if the sale was successful, false if the sale was unable to be
+	 *         processed. Note, a false return is indicitive of the quantity being too
+	 *         high for this company to support a sale. For instance, the quantity may
+	 *         cause the company to exceed its monthly sales potential, as specified by
+	 *         {@link #getMonthlySalePotential(Product)}
+	 */
+	public boolean registerSale(Product product, int quantity)
+	{
+		int salesInCurrentMonth = getThisMonthsSalesOf(product) + quantity;
+		
+		if (salesInCurrentMonth > getMonthlySalePotential(product))
+		{
+			return false;
+		}
+		else
+		{
+			float grossProfit = getRetailPriceOf(product) * quantity;
+			credit.processGrossProfit(grossProfit);
+			this.salesInCurrentMonth.put(product, salesInCurrentMonth);
+			return true;
+		}
+	}
+	
+	/**
+	 * If {@link #salesInCurrentMonth} does not contain the specified product, the product
+	 * will be added to the map and initialized with a value of 0, and that value will be
+	 * returned.
+	 * <p>
+	 * Otherwise, the corresponding value in {@link #salesInCurrentMonth} will be
+	 * returned.
+	 */
+	private int getThisMonthsSalesOf(Product product)
+	{
+		if (salesInCurrentMonth.containsKey(product))
+		{
+			return salesInCurrentMonth.get(product);
+		}
+		else
+		{
+			salesInCurrentMonth.put(product, 0);
+			return 0;
+		}
+	}
+	
+	/**
+	 * Equivalent to {@link #retailPrice}, but handles the case in which
+	 * {@link #retailPrice} is not registered for the given product.
+	 * <p>
+	 * As of version 0.1, all products must register a retail price, so an
+	 * {@link IllegalStateException} is thrown if {@link #retailPrice} does not contain
+	 * the given product. However, this functionality may be changed in future revisions
+	 * and is subject to change.
+	 */
+	private float getRetailPriceOf(Product product)
+	{
+		if (retailPrice.containsKey(product))
+		{
+			return retailPrice.get(product);
+		}
+		else
+		{
+			throw new IllegalStateException(
+					"Retail price must be registered for all products");
+		}
+	}
+	
+	/**
+	 * Register all expenses, process production (as specified by
+	 * {@link #processProduction()}), and update {@link #publicOpinionOfProducts} and
+	 * {@link #publicOpinionOfCompany} as well as all campaigns in
+	 * {@link #activeMarketing}
+	 */
+	public void processMonth()
+	{
+		processProduction();
+		float expense = getMonthlyEmployeeExpense() + getMonthlyMarketingExpense();
+		credit.processPayment(expense);
+		
+		List<MarketingCampaign> completedCampaigns = new LinkedList<>();
+		for (MarketingCampaign campaign : activeMarketing)
+		{
+			Product product = campaign.getTarget();
+			int publicityGain = campaign.processMonth();
```

Comment:
1: Good points. My assumption was that marketing campaigns can only lead to positive results, and any negative effects they had would be registered as an event (e.g. Your recent ad campaign was offensive and caused a drop in public opinion of your company/product, etc.).

1.1: PublicityGain in this context is the result of the campaign in the current month. While this may be influenced by factors such as the amount of publicity already obtained from the campaign and the length of the campaign, those factors are hidden from Company, so in this case it can be thought of as absolute (to be added to the current publicity). **I'll update the documentation of MarketingCampaign.processMonth()**

2: This is a style choice, that we should continue to discuss. Thank you for mentioning it.
I don't think **publicOpinionOfCompany += campaign.processMonth()** is as descriptive as **publicOpinionOfCompany += publicityGain**; the latter is significantly more readable, and thus more maintainable. It also helps when debugging, as you can watch the value of campaign.processMonth(), and better track exceptions, if they are to occur.

2.1: In most cases, you don't need to worry about the memory footprint of local variables, as the compiler will most likely remove them (so long as they are or can be properly scoped). Even if it didn't, Java's garbage collector is pretty good with local variables - enough that we don't need to nitpick over saving a byte or two. There is a decent page on [premature optimization](http://c2.com/cgi/wiki?PrematureOptimization) if you want some reading for your down time.